src/coreclr/jit/fgopt.cpp

```diff
@@ -317,9 +317,23 @@ PhaseStatus Compiler::fgPostImportationCleanup()
             {
                 LclVarDsc* returnSpillVarDsc = lvaGetDesc(lvaInlineeReturnSpillTemp);
 
-                if ((returnSpillVarDsc->lvType == TYP_REF) && returnSpillVarDsc->lvSingleDef)
+                if ((returnSpillVarDsc->lvType == TYP_REF))
                 {
-                    lvaUpdateClass(lvaInlineeReturnSpillTemp, retExprClassHnd, impInlineInfo->retExprClassHndIsExact);
+                    if (returnSpillVarDsc->lvSingleDef)
+                    {
+                        lvaUpdateClass(lvaInlineeReturnSpillTemp, retExprClassHnd,
+                                       impInlineInfo->retExprClassHndIsExact);
+                    }
+                    else if (lookupNamedIntrinsic(info.compMethodHnd) == NI_System_SZArrayHelper_GetEnumerator)
+                    {
+                        // If we are inlining System.SZArrayHelper.GetEnumerator, update the type of the return spill
+                        // temp regardless of whether it is single def or not.
+                        returnSpillVarDsc->lvClassHnd     = retExprClassHnd;
+                        returnSpillVarDsc->lvClassIsExact = impInlineInfo->retExprClassHndIsExact;
+#if DEBUG
+                        returnSpillVarDsc->lvClassInfoUpdated = true;
+#endif
+                    }
```

Comment:
I am not convinced this is correct. I think it would be better to leave a bool set when `lvaInlineeReturnSpillTemp` is created by `fgFindBasicBlocks` that it was a freshly created local for the returns of this function. When that bool is set you can update the class even if the local is not single-def.

Reply:
So if the spilling local is freshly created, I can update its class even for cases that are not `NI_System_SZArrayHelper_GetEnumerator`?

Reply:
I expect so, yes.

Reply:
Done.